packages/node/src/report.js

```diff
@@ -0,0 +1,224 @@
+/**
+ * Provides functions which report various types of _warnings_ to the user.
+ *
+ * @packageDocumentation
+ */
+
+import chalk from 'chalk'
+import { stripVTControlCharacters } from 'node:util'
+import { toKeypath } from 'to-keypath'
+import { InvalidArgumentsError } from './error.js'
+import { hrCode, hrLabel, hrPath } from './format.js'
+import { log as defaultLog } from './log.js'
+
+/**
+ * @import {ReportInvalidOverridesOptions, ReportSesViolationsOptions, StructuredViolation, StructuredViolationsResult} from './internal.js'
+ * @import {CanonicalName} from '@endo/compartment-mapper'
+ * @import {LavaMoatPolicy} from '@lavamoat/types'
+ */
+
+/**
+ * Default number of suggestions to make when reporting invalid canonical names
+ * in policy
+ */
+
+const DEFAULT_MAX_INVALID_CANONICAL_NAME_SUGGESTIONS = 3
+
+/**
+ * Gets the keypath for a canonical name within a LavaMoat policy object
+ *
+ * @param {LavaMoatPolicy} policy
+ * @param {CanonicalName} canonicalName
+ * @returns {string | undefined} The keypath as a string, or undefined if not
+ *   found
+ */
+const findCanonicalNameKeypath = (policy, canonicalName) => {
+  const { resources } = policy
+
+  // Check if it's a direct resource
+  if (canonicalName in resources) {
+    return toKeypath(['resources', canonicalName])
+  }
+
+  // Check nested packages
+  for (const [resourceName, resourcePolicy] of Object.entries(resources)) {
+    if (resourcePolicy.packages && canonicalName in resourcePolicy.packages) {
+      return toKeypath(['resources', resourceName, 'packages', canonicalName])
+    }
+  }
+
+  // Check include array
+  if (policy.include?.includes(canonicalName)) {
+    return toKeypath(['include', canonicalName])
+  }
+
+  return undefined
+}
+
+/**
+ * Reports resources from policy which weren't found on disk and are thus not in
+ * the compartment map descriptor.
+ *
+ * If no `policy` is provided, this function does nothing.
+ *
+ * @param {Set<CanonicalName>} unknownCanonicalNames - Set of canonical names
+ *   that were referenced in policy but not found
+ * @param {Set<CanonicalName>} knownCanonicalNames - Set of all canonical names
+ *   found in the compartment map
+ * @param {ReportInvalidOverridesOptions} options
+ * @returns {void}
+ * @internal
+ */
+export const reportInvalidCanonicalNames = (
```

Comment:
(not in this PR) Having this for webpack as well would be great. I'd like to migrate this to core or somewhere shared.